### PR TITLE
feat: python CdpEvmServerWalletProvider & CdpEvmSmartWalletProvider

### PR DESCRIPTION
**Pull request by CarsonRoscoe**
# CDP SDK v2 Integration

### Summary

This PR integrates CDP SDK v2 into our server and smart wallet providers, replacing the existing CDP v1 implementation. This update enables improved performance and access to new features while addressing challenges with Python dependency management and asynchronous operation.

#### CDP v2 Migration
* Replaced CDP SDK v1 with v2 throughout the codebase
* Updated wallet provider implementations to use the new CDP v2 API
* Asynchronous Operation Handling
* Added nest-asyncio to manage nested event loops in the OpenAI agents SDK framework
* Created utility methods to run async operations synchronously where needed

#### Testing Improvements
* Refactored tests to properly mock asynchronous CDP v2 methods
* Fixed issues with test assertions to verify results rather than method calls
* Enhanced error handling test coverage
* Updated address validation tests to reflect CDP v2's approach

#### Code Quality Enhancements
* Expanding linting, as only the `langchain` framework extension and a couple key examples were in the scope of the lint/format checks

#### Technical Details
* CDP v2 is an async-first library, requiring significant changes to our synchronous interfaces
* Used asyncio.run() and custom wrapper methods to bridge async/sync boundaries
* Fixed edge cases around error handling, particularly for invalid addresses
* Ensured consistent behavior across server and smart wallet providers

#### Testing

**Unit Tests**

Rewrote all CDP related unit tests. All pass, though there are some lingering warnings related to async work.

**Examples**

Tested every example manually

**CLI**

Generated langchain and openai agents sdk chatbots for both cdp server wallets and cdp smart wallets

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: python/coinbase-agentkit/coinbase_agentkit/action_providers/cdp/cdp_api_action_provider.py [hunk 1]
```diff
@@ -1,125 +1,120 @@
...
-- network: The network the address is on (e.g. "base-mainnet")
-- address: The Ethereum address to check
-
-Important notes:
-- This tool will not work on base-sepolia, you can default to using base-mainnet instead
-- The wallet's default address and its network may be used if not provided
-""",
-        schema=AddressReputationSchema,
+This tool will request test tokens from the faucet for the default address in the wallet. It takes the wallet and asset ID as input.
+Faucet is only allowed on 'base-sepolia' or 'solana-devnet'.
```
Comment (0xRAG) [inline python/coinbase-agentkit/coinbase_agentkit/action_providers/cdp/cdp_api_action_provider.py:49]:
```suggestion
Faucet is only allowed on 'base-sepolia', 'ethereum-sepolia' or 'solana-devnet'.
```

File: python/coinbase-agentkit/coinbase_agentkit/action_providers/cdp/cdp_api_action_provider.py [hunk 2]
```diff
@@ -1,125 +1,120 @@
...
+If fauceting on 'base-sepolia', user can only provide asset ID 'eth', 'usdc', 'eurc' or 'cbbtc', if no asset ID is provided, the faucet will default to 'eth'.
```
Comment (0xRAG) [inline python/coinbase-agentkit/coinbase_agentkit/action_providers/cdp/cdp_api_action_provider.py:50]:
```suggestion
If fauceting on 'base-sepolia' or 'ethereum-sepolia', user can only provide asset ID 'eth', 'usdc', 'eurc' or 'cbbtc', if no asset ID is provided, the faucet will default to 'eth'.
```

File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_server_wallet_provider.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+            self._web3 = Web3(Web3.HTTPProvider(rpc_url))
+
+            client = self.get_client()
+            if config.address:
+                account = asyncio.run(self._get_account(client, config.address))
+            else:
+                account = asyncio.run(self._create_account(client))
+
+            self._address = account.address
+            self._account = account
```
Comment (0xRAG) [inline python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_server_wallet_provider.py:78]: nit: you could probably reduce the state to just `self._account`, and get the address from `self._account.address` wherever you need it

File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_server_wallet_provider.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+        """
+        client = self.get_client()
+
+        async def _sign_typed_data():
+            async with client as cdp:
+                return await cdp.evm.sign_typed_data(
+                    address=self.get_address(),
+                    typed_data=typed_data,
+                )
```
Comment (0xRAG) [inline python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_server_wallet_provider.py:272]: There is no top-level `typed_data` parameter. See here for a usage: https://github.com/coinbase/cdp-sdk/blob/main/examples/python/evm/account.sign_typed_data.py

File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+
+        Returns:
+            Cdp: A new CDP client instance
+
+        """
+        return CdpClient(
+            api_key_id=self._api_key_id,
+            api_key_secret=self._api_key_secret,
+            wallet_secret=self._wallet_secret,
+            debugging=True,
```
Comment (0xRAG) [inline python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py:138]: did you mean to leave this?

File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+        Returns:
+            str: The transaction hash as a string
+
+        """
+        value_wei = Web3.to_wei(value, "ether")
+        client = self.get_client()
+
+        async def _send_user_operation():
+            async with client as cdp:
+                smart_account = await cdp.evm.get_smart_account(
```
Comment (0xRAG) [inline python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py:211]: any reason you're not also storing smart_account as a class variable?

File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_server_wallet_provider.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+
+        Args:
+            client (CdpClient): The CDP client instance
+
+        Returns:
+            Any: The newly created account object
+
+        """
+        async with client as cdp:
+            return await cdp.evm.create_account(idempotency_key=self._idempotency_key)
```
Comment (0xRAG) [inline python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_server_wallet_provider.py:327]: When I run the chatbot without setting an idempotency key, this line throws an error:
```
Starting Agent...
Traceback (most recent call last):
  File "/Users/ryan/code/public/agentkit/python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_server_wallet_provider.py", line 75, in __init__
    account = asyncio.run(self._create_account(client))
  File "/Users/ryan/code/public/cdp-sdk/python/.venv/lib/python3.10/site-packages/nest_asyncio.py", line 30, in run
    return loop.run_until_complete(task)
  File "/Users/ryan/code/public/cdp-sdk/python/.venv/lib/python3.10/site-packages/nest_asyncio.py", line 98, in run_until_complete
    return f.result()
  File "/Users/ryan/.pyenv/versions/3.10.16/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception.with_traceback(self._exception_tb)
  File "/Users/ryan/.pyenv/versions/3.10.16/lib/python3.10/asyncio/tasks.py", line 232, in __step
    result = coro.send(None)
  File "/Users/ryan/code/public/agentkit/python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_server_wallet_provider.py", line 327, in _create_account
    return await cdp.evm.create_account(idempotency_key=self._idempotency_key)
  File "/Users/ryan/code/public/cdp-sdk/python/cdp/analytics.py", line 114, in wrapper
    raise error
  File "/Users/ryan/code/public/cdp-sdk/python/cdp/analytics.py", line 99, in wrapper
    return await func(*args, **kwargs)
  File "/Users/ryan/code/public/cdp-sdk/python/cdp/evm_client.py", line 60, in create_account
    evm_account = await self.api_clients.evm_accounts.create_evm_account(
  File "/Users/ryan/code/public/cdp-sdk/python/.venv/lib/python3.10/site-packages/pydantic/_internal/_validate_call.py", line 33, in wrapper_function
    return await wrapper(*args, **kwargs)
  File "/Users/ryan/code/public/cdp-sdk/python/.venv/lib/python3.10/site-packages/pydantic/_internal/_validate_call.py", line 111, in __call__
    res = self.__pydantic_validator__.validate_python(pydantic_core.ArgsKwargs(args, kwargs))
pydantic_core._pydantic_core.ValidationError: 1 validation error for EVMAccountsApi.create_evm_account
x_idempotency_key
  String should have at least 36 characters [type=string_too_short, input_value='', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/string_too_short
```

Changing the line to `return await cdp.evm.create_account(idempotency_key=self._idempotency_key or None)` fixes it, so I think you'll want to fallback to `None` wherever you initialize the `self._idempotency_key` class variable

File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+            if not self._api_key_id or not self._api_key_secret or not self._wallet_secret:
+                raise ValueError(
+                    "Missing required environment variables. CDP_API_KEY_ID, CDP_API_KEY_SECRET, CDP_WALLET_SECRET are required."
+                )
+
+            if not owner_address_or_private_key:
+                raise ValueError("Owner private key or CDP server wallet address is required")
+
+            network_id = config.network_id or os.getenv("NETWORK_ID", "base-sepolia")
+            self._idempotency_key = config.idempotency_key or os.getenv("IDEMPOTENCY_KEY")
```
Comment (0xRAG) [inline python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py:68]: I don't think this is actually used anywhere in this class currently

**Comment by 0xRAG** [comment]
I'm seeing some strange behavior. Setup: langchain cdp server account chatbot, base-sepolia. Steps:
1. Requested eth from faucet
2. Send some eth to some address
3. Requested eth balance, and the reported balance is higher than before the send!

```
Prompt: send 0.00001 ETH to 0x6Cd01c0F55ce9E0Bf78f5E90f72b4345b16d515d

-------------------
Wallet Details:
- Provider: cdp_evm_server_wallet_provider
- Address: 0x3C0a13f6fAbcD3D4d1976eb112a31752c94FCcA9
- Network:
  * Protocol Family: evm
  * Network ID: base-sepolia
  * Chain ID: 84532
- Native Balance: 100000000000000
-------------------

-------------------
Successfully transferred 0.00001 native tokens to 0x6Cd01c0F55ce9E0Bf78f5E90f72b4345b16d515d.
Transaction hash: 0xac291246c0a6453af60face85f94f22ba5eb2fd7fe27398810cc0194b71e6e2d
-------------------
Successfully transferred 0.00001 ETH to the address **0x6Cd01c0F55ce9E0Bf78f5E90f72b4345b16d515d**. 

Transaction hash: **0xac291246c0a6453af60face85f94f22ba5eb2fd7fe27398810cc0194b71e6e2d**.
-------------------

Prompt: print wallet details
Here are your wallet details:

- **Wallet Address:** 0x3C0a13f6fAbcD3D4d1976eb112a31752c94FCcA9
- **Network ID:** base-sepolia
- **Chain ID:** 84532
- **Native Balance:** 0.0001 ETH (100000000000000 wei)
- **Provider:** cdp_evm_server_wallet_provider
-------------------

Prompt: get eth balance

-------------------
Native balance at address 0x3C0a13f6fAbcD3D4d1976eb112a31752c94FCcA9: 89978990066889
-------------------
Your current ETH balance is approximately **0.08997899 ETH** (or **89978990066889 wei**).
-------------------

Prompt: print wallet details
Here are your wallet details:

- **Wallet Address:** 0x3C0a13f6fAbcD3D4d1976eb112a31752c94FCcA9
- **Network ID:** base-sepolia
- **Chain ID:** 84532
- **Native Balance:** 0.08997899 ETH (89978990066889 wei)
- **Provider:** cdp_evm_server_wallet_provider
-------------------
```

File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_server_wallet_provider.py (hunk 3 above)
Comment (CarsonRoscoe) [reply python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_server_wallet_provider.py:78]: Checked and there were no references to `self._account`, just `self._address`, so removed that instead

File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py (hunk 5 above)
Comment (CarsonRoscoe) [reply python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py:138]: Nope

File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py (hunk 6 above)
Comment (CarsonRoscoe) [reply python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py:211]: I made an assumption that it would have the same socket issues as storing the CDP client. After testing, this was not true, so I am now storing the smart account directly

File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py (hunk 8 above)
Comment (CarsonRoscoe) [reply python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py:68]: You are right, it is not. There is currently no way to set a idempotency key when creating a smart wallet with the `cdp-sdk`, however I assume there will be in the future? If not, I'll delete this entirely

**Review by CarsonRoscoe — COMMENTED** [review]
> Requested eth balance, and the reported balance is higher than before the send!

@0xRAG Very weird indeed. I have seen the AI think that WEI to ETH is a 15 decimal difference rather than 18 many times. I don't think this is necessarily due to this PR, but will double check that our transfer & balance checks are consistent in their units.

**Review by CarsonRoscoe — COMMENTED** [review]
Confirmed our units are consistent in Python. We expect `ether` as our native_transfer param type, and `wei` as what our balance is stated in.

I think your bug was a mental brainfart from the AI seeing the balance in wei, and improperly moving the decimal over 18 places.

I'm been thinking lately that we probably want the wallet details to show both units, for the AI's sake. Give it more of a hint. However, that is not within the scope of this PR

**Review by 0xRAG — APPROVED** [review]
